### ISSUE TITLE
How to evaluate molecule generation models?

### ISSUE THREAD

**Issue by chaitjo** [DeepGraphLearning/torchdrug#46]
Hi torchdrug team, thank you for the awesome project! I am playing with molecule generation models, and am interested in trying to reproduce the benchmarks posted here: https://torchdrug.ai/docs/benchmark/generation.html

I am able to follow the tutorial for molecule generation: https://torchdrug.ai/docs/tutorials/generation.html

But I found that there was no mention of how we can evaluate models once they are fully trained. Is there any evaluator class or oracle that can be called to obtain the metrics as in your benchmark? 

Additionally, do you have any advice on how to set the hyperparameters to fairly reproduce/compare to the GCPN or GraphAF papers?

**Comment by chaitjo** [comment]
A quick suggestion (feel free to ignore): Maybe something that would be super beneficial to the community could be to release the scripts used to conduct the benchmarks (not just for molecule generation, but other tasks, too). Kind of similar to OGB examples, based on which many people have built up new models.

**Comment by chaitjo** [comment]
Bump, @KiddoZhu

**Comment by KiddoZhu** [comment]
Hi! I only have a rough idea of the benchmark of GCPN and GraphAF. The evaluation is just carried out during the finetuning stage, by picking the top-k molecules generated during the finetuning.

 @shichence is the first author of GraphAF and he may tell you more about the details.

**Comment by shichence** [comment]
Hi! For optimal hyperparameters, we have some example codes and configs,  and we plan to release them once they are beautified.
For evaluation, just as @KiddoZhu mentioned, we maintain a dictionary dynamically during finetuning, which contains the top-k molecules (and their scores) generated by the model. To obtain validity, novelty, uniqueness scores as reported in the original paper, you can first generate a set of molecules (e.g. 5000 molecules), and then calculate these metrics.

**Comment by chaitjo** [comment]
Hi @shichence @KiddoZhu, thank you for the response, I do look forward to the release of the example codes and configs! I think those are exactly what I need. 

I am guessing that the dictionary during finetuning will produce such an output, as in your tutorial: https://torchdrug.ai/docs/tutorials/generation.html#goal-directed-molecule-generation-with-reinforcement-learning-gcpn
```
The results are as follows:

(0.948, 'C1=CC=C(CNC2=NC=NCC3=CN2C(C2=COCC2)=C3)C=C1')
(0.948, 'CCC1=CC=CC=C1NC(=O)C12CC(=O)N(C1)C1=CC=CC=C12')
(0.947, 'O=C1CCNC(C2=CC=CN=C2)CN1CC1=CC=CC(Cl)=C1')
(0.947, 'CC1=C(C(=O)NC2CCCN(C3=CC=CC=C3)C2)C=CN=C1')
(0.947, 'CCNC1CCC2=CC=CC(=C2)N(C(=O)C2=CC=CC=N2)C1')
(0.946, 'O=C(C1=CC=CC=C1F)N1CC2=CC=CC=C2C(CCO)C1')
...
```

I wasn't able to find out from the tutorial how I may actually access this dictionary. On going through the code, is this the `self.best_results[task]` dictionary? 

If so, what function shall I call in order to compute the final metrics on top of this? E.g. can I use: https://github.com/DeepGraphLearning/torchdrug/blob/master/torchdrug/metrics/metric.py#L166?

**Comment by chaitjo** [comment]
Great, thank you very much!

**Comment by dgm2** [comment]
> For both stages, 10 epochs should be enough. For learning rate and other hyperparameters related to reinforcement learning, you can follow the molecule generation tutorials. The results obtained in the tutorial are very close to the results reported in the original paper.
> 
> Other observations:
> 
> * setting the `max_node` to a large number will make the penalized logp very high.

Hello, 
Anyone knows what would be the number for a "high" max_node e.g. to replicate GraphAF results in plopg ? 
(by default is set to 32)
Many thanks!
